include/envoy/http/header_map.h

```diff
@@ -659,6 +660,13 @@ class HeaderMap {
     headers.dumpState(os);
     return os;
   }
+
+  /**
+   * Return the optional stateful formatter attached to this header map. Filters can use the
+   * non-const version to remember additional header keys during operation if they wish.
+   */
+  virtual StatefulHeaderKeyFormatterOptConstRef formatter() const PURE;
+  virtual StatefulHeaderKeyFormatterOptRef formatter() PURE;
```

Comment:
In my admittedly abnormal use case, I need to set specifically-cased headers from a filter extension. Is the suggested way to set the map value on the header map as normal, then invoke `rememberOriginalHeaderKey` on this if non-null? (I can't think of any better way without scary additional overloads in HeaderMap, just confirming)

Reply:
Yeah I think that would be the best way. There are too many ways to set a cased string on headers today that I don't think providing another set of functions that preserve the cased value (which would just do this `if (formatter) { formatter->rememberOriginalKey(key) }` check anyways) is worth it.

Reply:
Yes that's the way. I will add more comments.

> How about handling this from Request and Response Headers to add logic within routes and virtual hosts?

Sorry what do you mean?

Reply:
Will reword. I was thinking about VirtualHosts, particularly pushed through VHDS, which may need to use PreservedCase for headers add via request_headers_to_add and response_headers_to_add.

Reply:
> Will reword. I was thinking about VirtualHosts, particularly pushed through VHDS, which may need to use PreservedCase for headers add via request_headers_to_add and response_headers_to_add.

Yeah that is not supported today, but I think could be done with a custom filter or possible as an addition to the router functionality. I will add comments around this and we can track as an optional follow up.